src/remark.1
.\"                                      -*- nroff -*-
.\" remark.1 - Manual page for remark.
.\"
.\" Copyright (C) 2001-2005 Oskar Liljeblad
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU Library General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software
.\" Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
.\"
.TH REMARK "1" "May 4, 2005" "remark (regex-markup)"
.SH NAME
remark \- Highlight text using regular expressions
.SH SYNOPSIS
.B remark
.RI [ OPTION ]... " " [ FILE ] " " [ TEXT ]...
.SH DESCRIPTION
This manual page document describes the \fBremark\fP command.

The \fBremark\fP program (regex-markup) performs regular expression-based
text markup according to used-defined rules. This can be used to color
syslog files as well as the output of programs such as ping,
traceroute, gcc etc. The hierarchal rules-format is simple to
understand, yet very powerful.

The \fBremark\fP program reads lines of text from standard in (or uses
TEXT as input string). Each line is highlighted according to rules
defined in FILE, and is then printed to standard out. Lines are
optinally wrapped (split) to fit within some specified width.
.SH OPTIONS
These programs follow the usual GNU command line syntax, with long
options starting with two dashes (`-').
.TP
.B \-p, \-\-prepend=\fISTRING\fR
When wrapping, this specifies a string that will be added
to the end of all wrapped lines (except the last one).

It is possible to highlight this string from rule scripts
by using the `prepend' keyword. See below for more information.
.TP
.B \-a, \-\-append=\fISTRING\fR
When wrapping, this specifies a string that will be added
before all wrapped lines (but after any retained text).
The first wrapped line will not have this string appended.

It is possible to highlight this string from rule scripts
by using the `append' keyword. See below for more information.
.TP
.B \-r, \-\-retain=\fICOUNT\fR
When wrapping, this allows a number of characters from the
beginning line to be copied to all wrapped lines, as a
common header.
.TP
.B \-w, \-\-width=\fICOLUMNS\fR
Specifies maximum width of printed lines. If this is specified
but not \-\-wrap, then character wrapping (\-\-wrap=char) is
automaticly enabled.
.TP
.B \-f, \-\-wrap=\fIMODE\fR
Specify wrapping type. The possible values for MODE are
`none' or `n' for no wrapping, `char' or `c' for wrapping that
breaks between character, and `word' or `w' for wrapping that
breaks between words.
.TP
.B \-\-help
Show summary of options.
.TP
.B \-\-version
Output version information and exit.
.SH RULE FORMAT
The remainder of this manual describes the format of highlighting rules
used by regex-markup. For the short version, look at the example files
in the examples subdirectory. The hierarchal format should be
mostly self explanatory.
.PP
A rule file consists of style and macro definitions, and match
statements. Order of match statements is important, since they
are executed from top down. Styles and macros need to be
defined before they are used.
.PP
The syntax of rule files is similar to that of programming
languages such as C and Java in that indentation and use of
whitespace doesn't matter. The following are all valid:
.PP
  style foobar { pre "blah" }

  style foobar
  {
  pre
  "blah"
  }

  style foobar pre "blah"
.PP
The braces ('{' and '}') have been skipped in the last example.
Similar to the programming languages, this allows only one
statement (`pre "blah"') to be associated with the style
(`style foobar').
.SH STYLES
A style defines what to do with a piece of text when some it has
been decided that it is highlighted. Each style has a unique
name which is case sensitive. Any character that doesn't interfere
with the parsing is allowed in style names. 

Since the \fBremark\fP program outputs to terminals, a style is basicly
just a string to insert before the style begins, and a style to
insert after. These are specified with `pre' and `post':

  style MyBoldStyle
  {
    pre "\\E[1m"
    post "\\E[0m"
  }

E.g. If a piece of text, say `foobar', was marked up with the style
above, the result would be `\\E[1mfoobar\\E[0m`.

It is also possible to include other styles in a style:

  style BoldBlue {
    pre "\\E[34m" post "\\E[39m"
    MyBoldStyle
  }
.SH MATCHING AND HIGHLIGHTING
By matching parts of text, those parts can be highlighted by the use of
styles. A match is written as a regular expression, same syntax as in
Perl. E.g.

  /foo/ MyBoldStyle
  /bar/ { BoldBlue }

(The braces in the second line are unnecessary and could be elimited.)

The first statement (/foo/ MyBoldStyle) would color the first occurrence
of `foo' in the line with the style MyBoldStyle. The second would color
the first occurrence of `bar' with BoldBlue.

Matches are hierarchal, which is why the following is possible:

  /foobar/ {
    MyBoldStyle
    /bar/ BoldBlue
  }

The first `foobar' occurrence found will be colored like this:
`foo' will be colored MyBoldStyle (actually `foobar' will),
and `bar' BoldBlue. The result is that `foobar' is bold, and
`bar' blue. There's no limit to the depth of hierarchal matches,
as long as there's free memory.

The \fBremark\fP program uses extended POSIX regular expressions (similar to
those used in Perl), which are described in the manpage regex(7).
The following would color `fo' followed by any number of `o' using
MyBoldStyle.

  /fo+/ MyBoldStyle

Note that these regex matches are case sensitive by default. This can
be changed by using the `i' option just after the last slash:

  /bar/i MyBoldStyle

This would match `bar', `BAR', `Bar' etc. Sometimes you also want to
highlight all occurrencies of a work. Do this with the `g' option:

  /bar/g MyBoldStyle

Remember that \fBremark\fP tries matches in one line at a time.
.SH SUBMATCHES
Submatches or subexpressions in regular expressions are used to selects
parts of the matched text of the regex. The following regex has two
subexpressions:

  /([0-9]+) seconds left, ([0-9]+) elapsed/

Subexpressions are numbered from 1 and up. To color only a particular
subexpression, write like this:

  /([0-9]+) seconds left, ([0-9]+) elapsed/ {
    red
    1 blue
    2 green
  }

In this case, the whole line would be colored red, with the exception of
the first and second subexpressions, which would be colored blue and green,
respectively. It is also possible to continue matching in subexpressions:

  /([0-9]+) seconds left, ([0-9]+) elapsed/ {
    red
    1 {
      blue
      /^0$/ red
    }
    2 green
  }
.SH MULTIPLE MATCHES
It is possible to execute the same statements for multiple matches or
subexpressions by separating the matches with a comma (`,'):

  /foo/,/bar/ blue

  /m(e)(e)k/ {
     1,2 blue
  }
.SH MACROS
A macro is a grouping of commands, which may be executed at any time.
Macros are similar to styles in that they are defined once and then
used in one or more places. 

The following macro colors the string it is passed blue, unless it is
`0', in which case it is colored red:

 macro digit-color {
   blue
   /^0$/ red
 }

This macro can be used to simplify the submatch example above:

  /([0-9]+) seconds left, ([0-9]+) elapsed/ {
    red
    1 digit-color
    2 green
  }
.SH SUBSTITUTION
Substitution - replacing text - is done with the s// construct:

  s/foo/bar/

This would replace the first occurrence of `foo' with `bar'.
Here /foo/ is a regular expression, and /bar/ is a special
replacement string.

The thing special with the replacement string is that it is
possible to reference subexpressions in the regex part
of the substitution. This is done with the dollar sign (`$'),
followed by the number of the subexpression.

  s/f(o+)/b$1/

This would replace `fo' to `bo', `foo' to `boo' etc.
It is also legal to write `${1}' instead of `$1' (this
might be necessary when the character following `$1' is
actually a digit).

The following dollar-variables may be used in the replacement
string:

$`	The string preceding what was matched

$&	The whole string that matched (also $0)

$'	The string following what was matched

The flags `i' and `g' described above may be used as well.

`set' is another substitution statement, implement to make
substitution more simple. It replaces the text in the current
context with an expanded string. Example:

  /foo/ set "bar"
  
This would cause the first occurence of "foo" to be replaced with "bar".
This is equivalent to:

  s/foo/bar/

However, in some situations `set' is easier to write:

  /complex match string/ { blue set "bar" }

rather than

  /complex match string/ { blue s/complex match string/bar/ }

The string specified as an argument to "set" is expanded according to
the rules for s// above, making this possible:

  /(foo)(bar)/ set "$2$1"

which would replace "foobar" with "barfoo". Note that $` and $' might
have no replacement (i.e. will be replaced by the empty string) when
used with `set'.
.SH CONTROL STATEMENTS
Two control statements are supported, `skip' and `break'.

Whenever a `skip' statement is encountered, the current line
will be discarded and not displayed at all. This can be useful
to hide unnecessary lines from the user. E.g.

  /CRON.*mail/ skip

The `break' statement can be used to speed up execution (i.e.
matching and highlighting). Whenever it is encountered, the
matching is considered to be finished, and the line will
be printed immediately (as marked up).
.SH PREPEND AND APPEND
It is possible to highlight the strings specified with
`--prepend' and `--append'. This is done by using the
`prepend' (or `pre') and `append' (or `post') keywords
like this:

  prepend { blue }

  append { green }

Of course, all statements described above may be used instead
of  `blue' and `green'.
.SH EXAMPLES
Highlight ping output using the ping rules file:
.br
	\fBping localhost | remark ../examples/ping.rules\fP
.SH REPORTING BUGS
Report bugs to <\fIoskar@osk.mine.nu\fP>.
.SH AUTHOR
\fBremark\fP was written by Oskar Liljeblad <\fIoskar@osk.mine.nu\fP>.
.SH COPYRIGHT
Copyright \(co 2001-2005 Oskar Liljeblad

This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.


NEWS
2005-05-22: regex-markup 0.10.0 released.
  Localization support added.
  Use gnulib for improved compatibility.
  Only use lstat64 when available (reported by Michal Fojtik).
  Tested on FreeBSD 4.10 (thanks to Michal Fojtik).
  Update manual pages, add copyright and bug reporting address.
  Cleaned up RPM package spec file.
  Added full GNU GPL header to source files and manual pages.
  Support --help and --version in cacheline.
  Lexer code now includes config.h before everything else.

2003-04-11: regex-markup 0.9.0 released.
  Implemented new command 'set', to simplify substitution constructs.
  Upgraded to autoconf 2.57 and automake 1.7.3.
  Fixed bug that would cause remark to die with "cannot allocate memory"
  when using $n where n is out of range in substitution rules.
  Added new testsuites bug05, e82, e83, e84, e85, e86.
  Clarified README about rule format location.
  Never ignore ctrl+c/SIGINT - I don't remember the benefit of ignoring it
  in the first place. If someone knows, please mail me. :)

2002-01-06: regex-markup 0.8.0 released.
  gcc 2.95 is now required to compile regex-markup.
  Fixed a bug with /(..)*/ matches that would cause regex-markup to
  highlight incorrectly or die ("Cannot allocate memory").
  Fixed a bug in s/../../ rules that would produce wrong output.
  Added example rules for make.
  Improved syslog highlighting rules.

2001-12-21: regex-markup 0.7.0 released.
  Added section about common problems to the README file.
  Remark now also looks in $HOME/.remark and /usr/share/regex-markup for
  rule files.
  The example rule files are installed in /usr/share/regex-markup by
  default.
  Added files for making an RPM package of regex-markup.

2001-12-19: regex-markup 0.6.0 released.
  Cleaned up code to reduce unnecessary memory allocations.
  Added new example rules for diff.
  Rewrote the cacheline program has in C.
  Fixed how escaped characters are dealt with in substitution strings.
  Improved terminal width detection in logwatcher.

2001-12-17: regex-markup 0.5.0 released.
  Almost all code has been rewritten and improved, resulting in more
  features, faster execution, cleaner and less buggy code.
  Implemented line wrapping in remark (previously this was done
  separately). As a result of this, foldlog was removed and logwatcher
  updated to use the build-in wrapping.
  Added a bunch of test cases (to a total of 558 tests).

2001-12-03: regex-markup 0.4.0 released.
  Began support for Cygwin by adding compatibility files from fileutils.
  Updated copyright.
  Added a new program "foldlog" to fold syslog lines.
  Added a testsuite for foldlog.
  Added missing file bison.simple.fixed (needed for complete rebuild).
  Improved error messages.
  Improved the syslog highlighter (logwatcher).
  Improved memory management (don't allocate and free as much memory).
  Fixed a bug in the parsing code that would cause freed memory to be read.

2001-10-28: regex-markup 0.3.0 released.
  Updated manual page.
  Fixed a substitution bug that would cause segmentation violation.
  Fixed style bugs causing incorrect highlighting.
  Better signal handling.
  Improved substitution with support for $`, $&, and $' variables.

2001-10-25: regex-markup 0.2.0 released.
  Added manual pages for remark.
  Added files for making a regex-markup Debian package.
  The author has changed email-address to oskar@osk.mine.nu.

2001-08-13: regex-markup 0.1.0 released.
  First public release.


src/remark.c
/* remark.c - Option-parsing and main function for regex-markup.
 *
 * Copyright (C) 2001-2005 Oskar Liljeblad
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Library General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

#if HAVE_CONFIG_H
#include <config.h>
#endif
/* POSIX */
#include <unistd.h>
/* C89 */
#include <signal.h>
#include <stdlib.h>
#include <string.h>
/* gnulib */
#include <getopt.h>
#include "dirname.h"
#include "progname.h"
#include "getline.h"
#include "version-etc.h"
/* Gettext */
#include <gettext.h>
#define _(String) gettext(String)
/* common */
#include "common/strbuf.h"
#include "common/string-utils.h"
#include "common/io-utils.h"
#include "common/error.h"
#include "common/llist.h"
#include "common/intutil.h"
/* regex-markup */
#include "remark.h"

#define PROGRAM "remark"

enum {
	VERSION_OPT	= 1000,
	HELP_OPT,
};

static uint32_t prepend_len = 0;
static uint32_t append_len = 0;
static char *prepend_string = "";
static char *append_string = "";

static char *short_opts = "a:f:p:r:w:";
static struct option long_opts[] = {
	{ "prepend",   required_argument, NULL, 'p' },
	{ "append",    required_argument, NULL, 'a' },
	{ "retain",    required_argument, NULL, 'r' },
	{ "width",     required_argument, NULL, 'w' },
	{ "wrap",      required_argument, NULL, 'f' },
	{ "version",   no_argument,	  NULL, VERSION_OPT },
	{ "help",      no_argument,	  NULL, HELP_OPT },
	{ 0 },
};

static void display_help(void);
static void display_version(void);

const char version_etc_copyright[] = "Copyright (C) 2004-2005 Oskar Liljeblad";

static void
display_version(void)
{
	version_etc(stdout, PROGRAM, PACKAGE, VERSION, "Oskar Liljeblad", NULL);
}

static void
display_help(void)
{
	printf(_("Usage: %s [OPTION]... FILE [TEXT]\n\
Read lines of text from standard in (or use TEXT as input string), highlight\n\
according to rules defined in FILE, and print to standard out.\n\
\n\
  -p, --prepend=STRING       string to prepend to all split lines (except last)\n\
  -a, --append=STRING        string to append to all split lines (except first)\n\
  -r, --retain=COUNT         copy characters from first line to wrapped ones\n\
  -w, --width=COLUMNS        wrapping width\n\
  -f, --wrap=TYPE            specifies wrapping type (word/char/none).\n\
      --help                 display this help and exit\n\
      --version              output version information and exit\n\
\n\
Report bugs to <%s>.\n"), program_name, PACKAGE_BUGREPORT);
}

static void
try_line(RemarkScript *script, RemarkInput *input, char *text)
{
	strbuf_set(input->mb.buffer, text);
	input->mb.bufferlen = strlen(text);
	strbuf_set(input->append_mb.buffer, append_string);
	input->append_mb.bufferlen = append_len;
	strbuf_set(input->prepend_mb.buffer, prepend_string);
	input->prepend_mb.bufferlen = prepend_len;

	if (execute_script(script, input)) {
		fputs(strbuf_buffer(input->mb.buffer), stdout);
		fflush(stdout);
	}
}

int
main(int argc, char **argv)
{
	struct sigaction action;
	RemarkInput input;
	int c;

	set_program_name(argv[0]);

	setlocale(LC_ALL, "");
	bindtextdomain(PACKAGE, LOCALEDIR);
	textdomain(PACKAGE);

	memset(&action, 0, sizeof(struct sigaction));
	/* 2003-04-11 : I don't know why this is good in the first place... */
	/*if (!isatty(STDIN_FILENO)) {
		action.sa_handler = SIG_IGN;
		action.sa_flags = SA_RESTART;
		if (sigaction(SIGINT, &action, NULL) < 0)
			die_errno(NULL);
	}*/

	while ((c = getopt_long (argc, argv, short_opts, long_opts, NULL)) != -1) {
		switch (c) {
		case 'p':
			prepend_string = optarg;
			break;
		case 'a':
			append_string = optarg;
			break;
		case 'r':
			if (!parse_uint32(optarg, &wrap_retain))
				die(_("invalid retain value: %s"), optarg);
			break;
		case 'w':
			if (!parse_uint32(optarg, &wrap_width))
				die(_("invalid width: %s"), optarg);
			break;
		case 'f':
			if (!identify_wrapper(optarg))
				die(_("invalid wrapping type: %s"), optarg);
			break;
		case VERSION_OPT:
			display_version();
			exit(0);
			break;
		case HELP_OPT:
			display_help();
			exit(0);
			break;
		case '?':
			exit(1);
		}
	}

	append_len = strlen(append_string);
	prepend_len = strlen(prepend_string);
	if (wrapper == 0)
		wrapper = WRAPPER_CHAR;
	if (wrap_retain + append_len + prepend_len >= wrap_width)
		die(_("retain, append and prepend too long"));

	init_input(&input);

	if (argc-optind >= 2) {
		RemarkScript *script = parse_script(argv[optind]);
		for (c = optind+1; c < argc; c++)
			try_line(script, &input, argv[c]);
		free_script(script);
	}
	else {
		RemarkScript *script = NULL;
		size_t linesize = 0;
		char *line = NULL;

		script = parse_script(argc > optind ? argv[optind] : NULL);
		while (getline(&line, &linesize, stdin) != -1) {
			chomp(line);
			try_line(script, &input, line);
		}
		free(line);
		free_script(script);

		if (ferror(stdin))
			die_errno(_("cannot read from standard in"));
	}

	free_input(&input);

	exit(0);
}
